File: backend/app/schemas/farm.py
from datetime import date, datetime
from pydantic import BaseModel, ConfigDict


class FarmBase(BaseModel):
    model_config = ConfigDict(extra="ignore")

    name: str
    description: str | None = None
    boundary_geojson: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    area_hectares: float | None = None
    crop_id: int | None = None
    sowing_date: date | None = None
    soil_type: str | None = None
    irrigation_type: str | None = None
    photo_url: str | None = None


class FarmCreate(FarmBase):
    pass


class Farm(FarmBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime


class FarmValidationResponse(BaseModel):
    crop: str | None = None
    soil_type: str | None = None
    suitability: str
    explanation: str
    recommended_action: str
    amendments_required: str | None = None
    irrigation_adjustment: str | None = None


class FarmCreateResponse(BaseModel):
    farm: Farm
    validation: FarmValidationResponse | None = None

File: frontend/src/types/farm.ts
export interface FarmValidation {
  crop: string | null
  soil_type: string | null
  suitability: string
  explanation: string
  recommended_action: string
  amendments_required?: string | null
  irrigation_adjustment?: string | null
}

export interface FarmCreateResponse {
  farm: Farm
  validation: FarmValidation | null
}

export interface Farm {
  id: number
  user_id: number
  name: string
  description?: string
  boundary_geojson?: string
  latitude?: number
  longitude?: number
  area_hectares?: number
  crop_id?: number
  sowing_date?: string
  soil_type?: string
  irrigation_type?: string
  photo_url?: string
  created_at: string
  updated_at: string
}

export interface FarmCreate {
  name: string
  description?: string
  boundary_geojson?: string
  latitude?: number
  longitude?: number
  area_hectares?: number
  crop_id?: number
  sowing_date?: string
  soil_type?: string
  irrigation_type?: string
  photo_url?: string
}

export interface Crop {
  id: number
  name: string
  scientific_name?: string
  growth_duration_days?: number
  description?: string
}

export interface CropGrowthStage {
  id: number
  crop_id: number
  stage_name: string
  min_day: number
  max_day: number
  notes?: string
}
